app/validation.go

```diff
@@ -48,28 +63,94 @@ func (f *Fissile) Validate() validation.ErrorList {
+		variableUsage[k.Name] = 0
+		if k.CVOptions.Internal {
+			// We always count internal variables as used
+			variableUsage[k.Name]++
+		}
+	}
+
+	for _, instanceGroup := range f.Manifest.InstanceGroups {
+		// Collect the names of properties used in this instance group
+		propertyDefaults := instanceGroup.CollectPropertyDefaults()
+		allErrs = append(allErrs, checkForUndefinedProperties(
+			fmt.Sprintf("instance_groups[%s].configuration.templates", instanceGroup.Name),
+			false,
+			instanceGroup.Configuration.Templates,
+			propertyDefaults)...)
+		allErrs = append(allErrs, checkForUndefinedVariables(
+			fmt.Sprintf("instance_groups[%s].configuration.templates", instanceGroup.Name),
+			instanceGroup.Configuration.Templates,
+			variableUsage,
+			false)...)
+		allErrs = append(allErrs, checkForSortedProperties(
+			fmt.Sprintf("instance_groups[%s].configuration.templates", instanceGroup.Name),
+			instanceGroup.Configuration.RawTemplates)...)
+	}
+
 	// All light opinions should differ from their defaults in the
 	// BOSH releases
-	allErrs = append(allErrs, f.checkLightDefaults(lightOpinions,
+	allErrs = append(allErrs, checkLightDefaults(lightOpinions,
 		boshPropertyDefaultsAndJobs)...)
 
+	allErrs = append(allErrs, checkNonConstantTemplates(f.Manifest.Configuration.RawTemplates)...)
+	allErrs = append(allErrs, checkForSortedVariables(f.Manifest.Variables)...)
+	allErrs = append(allErrs, checkForUndefinedVariables(
+		"configuration.templates",
+		f.Manifest.Configuration.Templates,
+		variableUsage,
+		true)...)
+	for variableName, variableUsageCount := range variableUsage {
+		if variableUsageCount == 0 {
+			allErrs = append(allErrs, validation.NotFound(
+				"variables",
+				fmt.Sprintf("No templates using '%s'", variableName)))
+		}
+	}
+
 	return allErrs
 }
 
-// Check that the given 'properties' are all defined in a 'bosh'
-// release.
-func checkForUndefinedBOSHProperties(label string, properties map[string]string, bosh propertyDefaults) validation.ErrorList {
-	// All provided properties must be defined in a BOSH release
+// checkForSortedProperties checks that the given ordered YAML map slice have
+// all of its keys in order.
+func checkForSortedProperties(label string, propertyOrder yaml.MapSlice) validation.ErrorList {
+	var previous string
 	allErrs := validation.ErrorList{}
 
-	for property := range properties {
+	for index, property := range propertyOrder {
+		key := property.Key.(string)
+		if index > 0 {
+			if key < previous {
+				allErrs = append(allErrs, validation.Forbidden(
+					fmt.Sprintf("%s[%s]", label, previous),
+					fmt.Sprintf("Template key does not sort before '%s'", key)))
+			}
+		}
+		previous = key
+	}
+
+	return allErrs
+}
+
+// checkForUndefinedProperties checks that all properties are known
+func checkForUndefinedProperties(label string, global bool, templates map[string]model.ConfigurationTemplate, knownProperties model.PropertyDefaults) validation.ErrorList {
```

Comment:
Here, we can make this function a method of a `validator` struct that takes the `knownProperties` as part of the initialization, and the `range templates` can happen on the caller side, so we don't need to pass the whole templates, neither the `global` parameter.

Reply:
Changed in 9b6136ede07c51872f28e995c24e15c78b4978df